relays/lib-substrate-relay/src/parachains/target.rs

```diff
@@ -115,15 +121,46 @@ where
 	async fn parachain_head(
 		&self,
 		at_block: HeaderIdOf<P::TargetChain>,
+		metrics: Option<&ParachainsLoopMetrics>,
 		para_id: ParaId,
 	) -> Result<Option<BestParaHeadHash>, Self::Error> {
-		let storage_key = parachain_head_storage_key_at_target(
+		let best_para_head_hash_key = best_parachain_head_hash_storage_key_at_target(
 			P::SourceRelayChain::PARACHAINS_FINALITY_PALLET_NAME,
 			para_id,
 		);
-		let para_head = self.client.storage_value(storage_key, Some(at_block.1)).await?;
+		let best_para_head_hash: Option<BestParaHeadHash> =
+			self.client.storage_value(best_para_head_hash_key, Some(at_block.1)).await?;
+		if let (Some(metrics), &Some(ref best_para_head_hash)) = (metrics, &best_para_head_hash) {
+			let imported_para_head_key = imported_parachain_head_storage_key_at_target(
+				P::SourceRelayChain::PARACHAINS_FINALITY_PALLET_NAME,
+				para_id,
+				best_para_head_hash.head_hash,
+			);
+			let imported_para_header = self
+				.client
+				.storage_value::<ParaHead>(imported_para_head_key, Some(at_block.1))
+				.await?
```

Comment:
Sorry, one more thing. `await?` returns the error directly, if any, while if `HeaderOf::<P::SourceParachain>::decode()` returns an error, we also log a message. Is this desired? Or should we log a message in both error cases ?

Reply:
If you're going to make a new method for double-encoded storage values, then let's log it here, where we know what the storage value actually mean and where we may log the meaningful error.

Ideally (imo) we'll need to migrate to some crate that would allow us to propagate error with context and log it once at the upper level (inside the relay loop). But that's not a part of your follow-up PR of course :)

Reply:
Sure, I will log the error here, but I was wondering if we should log an error both when `storage_value()` returns an `Err` and when `decode` returns an `Err`. Or should we keep the `await?` and log an error only when `decode()` returns an `Err` ?

The idea with having a new method for double-encoded storage value is not going very well so far. I think I might have to give it up.

Reply:
> Sure, I will log the error here, but I was wondering if we should log an error both when storage_value() returns an Err and when decode returns an Err. Or should we keep the await? and log an error only when decode() returns an Err ?

Imo more detailing logging = better :) The fact that logging occurs at the one branch only is probably the outcome of some bug that I've met. In the end the error is logged at the top level anyway, but this additional trace adds some context. So I'd add the second logging call here.